app/src/main/java/org/mozilla/focus/fragment/HomeFragment.kt

```diff
@@ -82,33 +109,76 @@ class HomeFragment : Fragment() {
+        unpackJsonHomeTiles(homeTiles, jsonArray)
+        val hta = HomeTileAdapter(onUrlEnteredListener, homeTiles)
+        adapter = hta
+        homeTileAdapter = hta
         layoutManager = GridLayoutManager(context, COL_COUNT)
         setHasFixedSize(true)
     }
 
+    private fun unpackJsonHomeTiles(homeTiles: MutableList<HomeTile>, homeTilesJsonArray: JSONArray) {
+        val homeTilesPrefs = context.getSharedPreferences(HOME_TILES_PREFS, MODE_PRIVATE)
+        val blacklist = homeTilesPrefs.getStringSet(bundled_blacklist, null)
+        for (i in 0..(homeTilesJsonArray.length() - 1)) {
+            val jsonObject = homeTilesJsonArray.getJSONObject(i)
+            val id = jsonObject.getString("id")
+            if (blacklist == null || !blacklist.contains(id)) {
+                homeTiles.add(BundledHomeTile.fromJSONObject(jsonObject))
+            }
+        }
+    }
+
     private fun initUrlInputView() = with (urlInputView) {
         setOnCommitListener {
             onUrlEnteredListener.onTextInputUrlEntered(text.toString(), urlInputView.lastAutocompleteResult, UrlTextInputLocation.HOME)
         }
         setOnFilterListener { searchText, view -> urlAutoCompleteFilter.onFilter(searchText, view) }
     }
 
+    fun addBundledTileToBlacklist(id: String) {
+        val blacklist = context.getSharedPreferences(HOME_TILES_PREFS, MODE_PRIVATE).getStringSet(bundled_blacklist, mutableSetOf())
+        val edit = context.getSharedPreferences(HOME_TILES_PREFS, MODE_PRIVATE).edit()
+        val newBlacklist = blacklist.toMutableSet()
+        newBlacklist.add(id)
+        edit.putStringSet(bundled_blacklist, newBlacklist)
+        edit.apply()
+        homeTileAdapter.removeItemAtPosition(getFocusedTile())
+    }
+
     companion object {
         const val FRAGMENT_TAG = "home"
 
         @JvmStatic
         fun create() = HomeFragment()
     }
+
+    fun getFocusedTile(): Int {
+        return (activity.currentFocus.parent as? RecyclerView)?.getChildAdapterPosition(activity.currentFocus) ?: RecyclerView.NO_POSITION
+    }
+
+    fun dispatchKeyEvent(event: KeyEvent): Boolean {
+        /**
+         * Key handling order:
+         * - Menu to control overlay
+         * - Return false, as unhandled
+         */
+        return handleSpecialKeyEvent(event)
+    }
+
+    private fun handleSpecialKeyEvent(event: KeyEvent): Boolean {
+        // Add code about handling home context menu
+        if (event.keyCode == KeyEvent.KEYCODE_MENU && getFocusedTile() != RecyclerView.NO_POSITION) {
+            activity.openContextMenu(view)
+            return true
+        }
+        return false
+    }
 }
 
-private class HomeTileAdapter(val onUrlEnteredListener: OnUrlEnteredListener, val tiles: List<HomeTile>) :
+private class HomeTileAdapter(val onUrlEnteredListener: OnUrlEnteredListener, val tiles: MutableList<HomeTile>) :
         RecyclerView.Adapter<TileViewHolder>() {
+    var currentFocusedIndex: Int = -1
```

Comment:
We talked about moving this out before - if it doesn't happen for this PR, file an issue with everything that needs to be cleaned up.